### ISSUE TITLE
Stable release of 1.1.0

### ISSUE THREAD

**Issue by akshayka** [cvxgrp/cvxpy#913]
This is a tracking issue for making 1.1.0 stable. We'll have at least one more prerelease, 1.1.0a2, before we get there.

Below is an incomplete outline of the release notes we might publish alongside the stable release.

New features
- Disciplined parametrized programming (for cone programs)
- Differentiation through cone programs (specifically, those solved by SCS)

Enhancements
- Allow use of the Python builtin `sum` function in DGP
- Allow use of cp.max, cp.min in DQCP (in addition to cp.maximum, cp.minimum)
- New atoms: support function

Bug fixes
- Dual variable recovery
- Detection of monotone functions for DQCP

Breaking changes
- Incompatible with Python 2
- Removed solvers: SuperSCS

Known issues
- Problems with many parameters can take a long time to compile

**Comment by bstellato** [comment]
It would be nice to have the QP canonicalization for DPPs in the stable release. I believe OSQP differentiation should be ready fairly soon. What do you think?

**Comment by akshayka** [comment]
It would indeed be great to have QP canonicalization supported in the stable release. It's awesome that the derivative is nearly ready.

I personally might not have the bandwidth to rewrite the QP canonicalization in the very near future; I'm not familiar with the code path, so it might take me some time. One option is to release a stable version soon, without QP derivatives, and add QPs when the code path is ready, in a 1.1.1 release.

On the other hand --- @SteveDiamond , you've mentioned before that getting QPs to work should be simple. Do you have the bandwidth to give it a shot soon, or should we release a stable version without it?

**Comment by SteveDiamond** [comment]
Yes, getting the QP canonicalization in place is something I can do. I'm going to China on January 15, so I'll aim to get it done before then.

Also, let's be sure to test all the solvers (including the various commercial ones) before we release the stable 1.1. Maybe I'll need to update the NAG interface, since I haven't heard anything from the developers.

**Comment by rostyboost** [comment]
> Problems with many parameters can take a long time to compile

I can confirm a pretty significant regression compared to `1.0.25` for the MIP problems we build. Is there someone working on this already? Otherwise happy to take a look.

**Comment by akshayka** [comment]
@rostyboost , no one is currently looking into this. You’re welcome to take a look!

**Comment by rileyjmurray** [comment]
Once PR #910 is merged, we'll also need web-documentation for the ``scalar_product`` atom.

Under "known issues", include: NAG interface doesn't work, and GUROBI conic interface doesn't work (GUROBI QP interface is fine). I think fixing GUROBI before a public ``1.1.0`` release is important.

**Comment by rileyjmurray** [comment]
Another known issue: ECOS-BB sometimes produces incorrect results. This was mentioned in (now closed) #578 and (still open) #631.

Something we might want to add: a deprecation warning against using the ``*`` operator for matrix multiplication (matrix-scalar and vector-scalar multiplication are fine). This was mentioned in #885.

**Comment by SteveDiamond** [comment]
What exactly can we do about ECOS-BB though? There's no other MIP solver that's easy to install (maybe the trial version of CPLEX actually). Not having a default MIP solver seems like a bad idea.

**Comment by rileyjmurray** [comment]
The ECOS-BB issue is a tough one. A short-term or medium-term solution would be to have cvxpy print a warning if ECOS-BB is used. The warning could (1) say that ECOS-BB has some unresolved bugs, and (2) recommend alternatives. The precise recommendation would take some thinking.

**Comment by rostyboost** [comment]
> and GUROBI conic interface doesn't work (GUROBI QP interface is fine). I think fixing GUROBI before a public `1.1.0` release is important.

How about dropping `gurobi_conif.py` and defaulting to `gurobi_qpif.py`? Potential benefits:
- from this backend API perspective, the distinction is arbitrary 
- `gurobi_qpif.py` implementation is much more succinct
- `gurobi_qpif.py` is likely a lot more efficient for socp problems as well, given that it's using some of the latest vectorized API calls of the Gurobi API (Gurobi >= 8.1.1)
- `gurobi_conif.py` implementation is messy. Why does the class extends from `SCS`?
-  avoid code duplication

**Comment by rileyjmurray** [comment]
> * from this backend API perspective, the distinction is arbitrary
> * `gurobi_qpif.py` is likely a lot more efficient for socp problems as well, given that it's using some of the latest vectorized API calls of the Gurobi API (Gurobi >= 8.1.1)

I don't quite follow. Do you mean that GUROBI can mix QP and SOCP models? Not all SOCP's can be written as QPs. If GUROBI is transforming QP constraints into SOCP constraints on the backend, then we might as well only have a conic interface.

**Comment by akshayka** [comment]
Now that #910 has been merged, I'd like to push a new alpha version, 1.1.0.a2. @rileyjmurray , @SteveDiamond , does that sound good to you guys?

**Comment by rileyjmurray** [comment]
Sounds good to me!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by akshayka** [comment]
Great, done, should be released on PyPi and the other channels soon. 

https://github.com/cvxgrp/cvxpy/releases/tag/v1.1.0a2

**Comment by akshayka** [comment]
Since the alpha release has been out for some time, and since we haven't pulled in bug fixes made on alpha to stable, we should try to make 1.1 stable soon.

@bstellato , what's the status of OSQP differentation?

@SteveDiamond , do you still want to implement DPP and canonicalization for QPs?

@rostyboost , any chance you had time to investigate the performance regressions you saw?

**Comment by SteveDiamond** [comment]
@akshayka I haven't been feeling well, so I can't promise to finish DPP and QP canonicalization in any reasonable time. I can help do a check of the remaining solvers though.

**Comment by SteveDiamond** [comment]
Could we push this week for getting a new version of the alpha on pip? I'm working now on the NAG interface. Some of the Gurobi tests fail on my machine, so we need to at least fix those.

**Comment by SteveDiamond** [comment]
I've removed the NAG interface. I spent a while trying to get it to work, but ultimately taking the long view I don't think it's a good idea to keep it. NAG doesn't have academic licenses, so it's going to be very difficult to maintain going forward. A while ago, I invited the original contributor of the NAG interface to help update it for cvxpy 1.1 and he did not respond.

**Comment by bstellato** [comment]
@akshayka I am really sorry but last month I did not have time to finalize the derivatives in OSQP because of teaching and other revisions. If you want something quickly, I can finish up the implementation as python functions in the osqp-python interface. However, for the C implementation it should take longer.

@SteveDiamond what do you think is required for the QP differentiation? From what I understood, we should change [qp_matrix_stuffing.py](https://github.com/cvxgrp/cvxpy/blob/master/cvxpy/reductions/qp2quad_form/qp_matrix_stuffing.py) to add the `ParamQuadProg` class in order to match [cone_matrix_stuffing.py](https://github.com/cvxgrp/cvxpy/blob/master/cvxpy/reductions/dcp2cone/cone_matrix_stuffing.py). Is there anything else apart from that? I might be able to help in a couple of weeks but at the moment the changes in `cone_matrix_stuffing.py` are quite obscure to me.

**Comment by SteveDiamond** [comment]
@bstellato that's the general idea. The real sticking point is this [gem of code](https://github.com/cvxgrp/cvxpy/blob/56c8ef6eb8e0529abc7c73d249db5c81850d8cba/cvxpy/utilities/coeff_extractor.py#L142). It's one of the nastiest parts of the whole codebase, and we need to figure out how it interacts with the new parameterized cvxcore. If you haven't finished with the OSQP C implementation, I think it's best to wait.

**Comment by rileyjmurray** [comment]
The contributing guide https://github.com/cvxgrp/cvxpy/blob/master/CONTRIBUTING.md should be updated to contain remarks about solver interfaces. For example:
* We doesn't ship major releases with broken solver interfaces. If a change to cvxpy breaks a solver interface and no one steps up to fix it, there is a chance that the interface could be removed.
* The core development team is generally willing to spend some effort on fixing a solver interface before removing it. However if we encounter a notable obstacle, the solver will likely be dropped. Examples of notable obstacles include:
  * The solver doesn't have an interface for a sufficiently high version of python. SuperSCS was dropped because it only had a python 2.7 interface, and cvxpy 1.1 requires python >= 3.5
  * The solver is commercial and doesn't have academic licenses. NAG was removed because it fell into this category, and no one stepped up to help the cvxpy development team.
  * The solver has been deprecated by its principal developer. Elemental fell into this category.

**Comment by bstellato** [comment]
> @bstellato that's the general idea. The real sticking point is this [gem of code](https://github.com/cvxgrp/cvxpy/blob/56c8ef6eb8e0529abc7c73d249db5c81850d8cba/cvxpy/utilities/coeff_extractor.py#L142). It's one of the nastiest parts of the whole codebase, and we need to figure out how it interacts with the new parameterized cvxcore. If you haven't finished with the OSQP C implementation, I think it's best to wait.

@SteveDiamond Thanks for the information. After solving a sequence of parametric QPs with release 1.1, I have realized it would be very beneficial to have `ParamConeProg` in place not only for differentation but also in terms of speedups. I would be happy to take a look at it but I would need a bit of help, maybe over Skype if you guys want to discuss it.

**Comment by rileyjmurray** [comment]
I’d like to add a fix for overflows in cvxcore for the 1.1 todos. Issues #958 #826 and #934 are affected by this. Issue #826 shows this happens on Windows and Linux workstations.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by andrea83ny** [comment]
Hello team, apologize if I write in here, but I see you have listed a few great enhancements and I thought to share a suggestion.
My knowledge of the math behing cvxpy is very limited so please ignore my thought if it doesn't make a lot of sense.
The simple suggestion is to enhance the verbose feature and make sure it prints to the screen more "user-friendly" messages for people like me who relies on the library without fully understanding what's happening when it's running.

**Comment by akshayka** [comment]
Adding more user-friendly error messages is a great idea, and something that we plan to do eventually. It's not in scope for 1.1. That said, we'll still be writing code and making new releases after 1.1 is made stable :)

**Comment by akshayka** [comment]
For some additional context, 1.1 is mostly about landing two new features, efficient re-compilation of parametrized programs and differentiating through solution maps of solvers.

**Comment by akshayka** [comment]
Here's one more action item (for me): DPP should compose with DGP and DQCP, even if the composition, for now, is just 'DPP isn't supported for these problem classes'.

**Comment by Alpima-Quant** [comment]
hi guys, any word on a release date for 1.1?

**Comment by rileyjmurray** [comment]
Resolved by https://github.com/cvxgrp/cvxpy/commit/d63bfa695643ecaadff6d8d0c4f8cf19a7ad0245.
